hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancer.java

```diff
@@ -417,24 +425,32 @@ private IterationResult doIteration() {
           findSourceStrategy.removeCandidateSourceDataNode(source);
         }
       }
-
-      if (!isMoveGenerated) {
-        //no move option is generated, so the cluster can not be
-        //balanced any more, just stop iteration and exit
-        return IterationResult.CAN_NOT_BALANCE_ANY_MORE;
-      }
-      return IterationResult.ITERATION_COMPLETED;
     } finally {
-      checkIterationMoveResults(selectedTargets);
```

Comment:
I think finally block better fits the idea of cleaning code, the new version seems to have too much process logic in it. Personally, I do like the original method.

Reply:
Yeah, but the original method logic is buggy. I think we can remove the finally block entirely if we don't return early from the `doIteration` method. If we have instead just break out from the while loop and return only at the end after processing results, we don't need try-finally.

Reply:
I agree, the ITERATION_RESULT is a little confused here. It's now referring to the 
1. reason for stop generating moves (MAX_SIZE_TO_MOVE_REACHED)
2. the result of inner iteration (ITERATION_COMPLETED)
3. the reason we should stop the whole balancer service (CAN_NOT_BALANCE_ANY_MORE)

Maybe we should sort out the process and refactor the code.